libcnb/src/layer/handling.rs

```diff
@@ -0,0 +1,686 @@
+use crate::Buildpack;
+use crate::{write_toml_file, TomlFileError};
+use serde::de::DeserializeOwned;
+use serde::Serialize;
+use std::fs;
+use std::path::Path;
+
+pub(crate) fn handle_layer<B: Buildpack + ?Sized, L: Layer<Buildpack = B>>(
+    context: &BuildContext<B>,
+    layer_name: impl AsRef<str>,
+    layer: L,
+) -> Result<LayerData<L::Metadata>, HandleLayerErrorOrBuildpackError<B::Error>> {
+    match read_layer(&context.layers_dir, layer_name.as_ref()) {
+        Ok(None) => handle_create_layer(context, layer_name, layer),
+        Ok(Some(layer_data)) => {
+            let layer_should_be_recreated = layer
+                .should_be_recreated(context, &layer_data)
+                .map_err(HandleLayerErrorOrBuildpackError::BuildpackError)?;
+
+            let layer_should_be_updated = layer
+                .should_be_updated(context, &layer_data)
+                .map_err(HandleLayerErrorOrBuildpackError::BuildpackError)?;
+
+            if layer_should_be_recreated {
+                delete_layer(&context.layers_dir, layer_name.as_ref())?;
+                handle_create_layer(context, layer_name, layer)
+            } else if layer_should_be_updated {
+                handle_update_layer(context, layer_data, layer)
+            } else {
+                // We need to rewrite the metadata even if we just want to keep the layer around
+                // since cached layers are restored without their types, causing the layer to be
+                // discarded.
+                write_layer(
+                    &context.layers_dir,
+                    &layer_data.name,
+                    layer_data.env,
+                    LayerContentMetadata {
+                        // We cannot copy the types from layer_data due to an issue with the current
+                        // libcnb implementation. The types will be missing in the TOML file on disk
+                        // but if they're not there, their default values will be used when
+                        // deserializing. Issue: https://github.com/Malax/libcnb.rs/issues/146
+                        //
+                        // Even if the deserialization of LayerContentMetadata is fixed, it would
+                        // not contain the layer types as they're not restored by the CNB lifecycle.
+                        // We must call layer.types here to get the correct types for the layer.
+                        types: layer.types(),
+                        metadata: layer_data.content_metadata.metadata,
+                    },
+                )?;
+
+                // Reread the layer from disk to ensure the returned layer data accurately reflects
+                // the state on disk after we messed with it.
+                read_layer(&context.layers_dir, layer_name.as_ref())?
+                    .ok_or(HandleLayerError::UnexpectedMissingLayer)
+                    .map_err(HandleLayerErrorOrBuildpackError::HandleLayerError)
```

Comment:
I can do that, I do think it's worth noting it's only used once whereas the the others are used in multiple locations. If we want to split this out, we probably want to do the same for the metadata recovery arm just below as well?

Reply:
Ah I missed that the others are used more than once. I agree that if we were to split this out, it would make sense for the metadata recovery arm too.

I'll leave it to you as to whether we should do this or not :-)

Reply:
I think I prefer it to stay the way it is for now. 🤔